### PR TITLE
fix(sdk): Polish the downloader/manager package error return

### LINKED ISSUE #8471: The sdk dependency manager error refers to Helm CLI
Output of `helm version`: v3.4.0 development and previous v3 releases

The download manager within the SDK notes... https://github.com/helm/helm/blob/6a536661fcf1dc0315798ece70bbf193e725945c/pkg/downloader/manager.go#L414

The SDK should have an error that's general and the CLI code in the cmd directory should detect it and decorate with any CLI related elements. Separation of concerns and all.

### PR DESCRIPTION
**Pull request by donggangcj**
Close #8471

Signed-off-by: Dong Gang <dong.gang@daocloud.io>

**What this PR does / why we need it**:
fix #8471
**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [X] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by donggangcj** [comment]
@mattfarina  I'm not sure if these modifications are consistent with your ideas.

**Review by jdolitsky — APPROVED** [review]

File: cmd/helm/dependency_build.go [hunk 1]
```diff
@@ -65,7 +67,11 @@ func newDependencyBuildCmd(out io.Writer) *cobra.Command {
 			if client.Verify {
 				man.Verify = downloader.VerifyIfPossible
 			}
-			return man.Build()
+			err := man.Build()
+			if e, ok := err.(downloader.ErrRepoNotFound); ok {
+				return errors.Errorf("no repository definition for %s. Please add the missing repos via 'helm repo add'", strings.Join(e.Repos, ", "))
```
Comment (bacongobbler) [inline cmd/helm/dependency_build.go:72]: couldn't this display repositories that might've succeeded? If I'm reading the code correctly, `e.Repos` and `missing` might not evaluate to the same result. There may only be a few repositories missing from the list.

I'd suggest appending the helpful error message to the output of `e.Error()`, like so:

```
return fmt.Errorf("%s. Please add the missing repos via 'helm repo add'", e.Error())
```

That way we leave the output of the error message up to the download manager.

**Review by bacongobbler — CHANGES_REQUESTED** [review]
Please write a unit test to ensure the output of `Error()` works as expected. Thanks.

File: cmd/helm/dependency_build.go (hunk 1 above)
Comment (donggangcj) [reply cmd/helm/dependency_build.go:72]: > couldn't this display repositories that might've succeeded? 

`Manager` method `hasAllRepos` ensures that all of the referenced deps are in the local repo cache and it doesn't return succeeded repositories. 

> If I'm reading the code correctly, `e.Repos` and `missing` might not evaluate to the same result. There may only be a few repositories missing from the list.

My bad! I did not process erros in the function `resolveRepoNames`，but i didn't find the reason why `e.Repos` and `missing` is different. Do you mean to merge `resolveRepoNames` with `hasAllRepos`?🤣

> 
> I'd suggest appending the helpful error message to the output of `e.Error()`, like so:
> 
> ```
> return fmt.Errorf("%s. Please add the missing repos via 'helm repo add'", e.Error())
> ```

Done; I removed some tips of `resolveRepoName` and is it necessary to show it here by a error handling function.

https://github.com/helm/helm/blob/241785c70fb38b2c074d7b3ddf0925812fb3fc69/pkg/downloader/manager.go#L469-L478

There seems to be an obvious error in this code

File: pkg/downloader/manager_test.go [hunk 2]
```diff
@@ -360,3 +360,32 @@ func TestBuild_WithRepositoryAlias(t *testing.T) {
...
+		name   string
+		fields fields
+		want   string
+	}{
+		{
+			name: "OK",
+			fields: fields{
+				Repos: []string{"https://charts1.example.com", "https://charts2.example.com"},
+			},
+			want: "no repository definition for https://charts1.example.com, https://charts2.example.com",
```
Comment (bacongobbler) [inline pkg/downloader/manager_test.go:378]: Can you please also add a test case for the issue we discussed earlier, where only a few repositories in the list are missing.

**Review by bacongobbler — APPROVED** [review]

**Pull request merged**